src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,92 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    private int size;
 
     @Override
     public void add(T value) {
-
+        if (elementData.length == size) {
+            elementData = grow();
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bound");
+        }
+        if (elementData.length == size) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
 
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index);
```

Comment:
```suggestion
        System.arraycopy(elementData, index + 1, elementData, index, // are you sure here?);
```

Reply:
Is it mistake ?
because it works without problems

Reply:
Is it mistake ?
because it works without problems ,or should i add " size - index - 1" ?

Reply:
it is ok both ways